pkg/app/server/grpcapi/piped_api.go

```diff
@@ -356,33 +356,63 @@ func (a *PipedAPI) ListNotCompletedDeployments(ctx context.Context, req *pipedse
 		return nil, err
 	}
 
-	opts := datastore.ListOptions{
-		Filters: []datastore.ListFilter{
-			{
-				Field:    "PipedId",
-				Operator: datastore.OperatorEqual,
-				Value:    pipedID,
-			},
-			// TODO: Change to simple conditional clause without using OR clause for portability
-			// Note: firestore does not support OR operator.
-			// See more: https://firebase.google.com/docs/firestore/query-data/queries?hl=en
-			{
-				Field:    "Status",
-				Operator: datastore.OperatorIn,
-				Value:    model.GetNotCompletedDeploymentStatuses(),
-			},
-		},
-	}
-
-	deployments, cursor, err := a.deploymentStore.List(ctx, opts)
-	if err != nil {
-		return nil, gRPCStoreError(err, fmt.Sprintf("list deployments of piped %s", pipedID))
-	}
-
-	return &pipedservice.ListNotCompletedDeploymentsResponse{
-		Deployments: deployments,
-		Cursor:      cursor,
-	}, nil
+    // Build filters for uncompleted deployments managed by this piped.
+    filters := []datastore.ListFilter{
+        {
+            Field:    "PipedId",
+            Operator: datastore.OperatorEqual,
+            Value:    pipedID,
+        },
+        // TODO: Change to simple conditional clause without using OR clause for portability
+        // Note: firestore does not support OR operator.
+        // See more: https://firebase.google.com/docs/firestore/query-data/queries?hl=en
+        {
+            Field:    "Status",
+            Operator: datastore.OperatorIn,
+            Value:    model.GetNotCompletedDeploymentStatuses(),
+        },
+    }
+
+    // Use deterministic ordering for stable pagination.
+    orders := []datastore.Order{
+        {
+            Field:     "UpdatedAt",
+            Direction: datastore.Desc,
+        },
+        {
+            Field:     "Id",
+            Direction: datastore.Asc,
+        },
+    }
```

Comment:
Could you test this on your local control-plane. Since we're using virtual column MySQL or FireStore as the database for controlplane, both require indexes to make queries work.